### PR TITLE
feat: add plugin_metadata into control API

### PR DESCRIPTION
**Pull request by kingluo**
### Description

Add plugin_metadata into control API, so that you could retrieve all plugin metadatas or specific plugin metadata from control API.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

@spacewander

### CODE REVIEW

File: apisix/control/v1.lua [hunk 1]
```diff
@@ -276,6 +277,28 @@ function _M.dump_service_info()
     return 200, info
 end
 
+function _M.dump_all_plugin_metadata()
+    local names = plugins.get_plugins_list()
```
Comment (spacewander) [inline apisix/control/v1.lua:281]: Better to use another way to do it. Maybe a new method in the `plugin` module or some other way.
We can't let the module in control-api require the module in admin-api. They are at the same level.

File: docs/en/latest/control-api.md [hunk 2]
```diff
@@ -451,3 +451,42 @@ Dumps the Upstream with the specified `upstream_id`:
    "modifiedIndex":1225
 }
 ```
+
+
+### GET /v1/plugin_metadatas
+
+Introduced in [v2.15](https://github.com/apache/apisix/releases/tag/2.15).
```
Comment (spacewander) [inline docs/en/latest/control-api.md:458]: Should be `2.15.0`. We have a patch version for the latest releases.

File: t/control/plugin-metadata.t [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+use t::APISIX 'no_plan';
+
+repeat_each(1);
+no_long_string();
+no_root_location();
+no_shuffle();
+log_level("info");
```
Comment (spacewander) [inline t/control/plugin-metadata.t:23]: We don't need to repeat the settings in https://github.com/apache/apisix/blob/44d20c391767d3d42c5111ea7ef6318e84633996/t/APISIX.pm#L23-L28

File: t/control/plugin-metadata.t [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            local _, body = t('/apisix/admin/plugin_metadata/file-logger',
+                ngx.HTTP_PUT,
+                [[
+                {"log_format": {"upstream_response_time": "$upstream_response_time"}}
+                ]]
+                )
+            ngx.say(body)
+        }
+    }
+--- request
```
Comment (spacewander) [inline t/control/plugin-metadata.t:62]: We can set the repeated sections in `add_block_preprocessor` like the other test files:
https://github.com/apache/apisix/blob/44d20c391767d3d42c5111ea7ef6318e84633996/t/deployment/conf_server.t#L28

File: t/control/plugin-metadata.t [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local _, body = t('/apisix/admin/plugin_metadata/example-plugin',
+                ngx.HTTP_PUT,
+                [[{
+                    "skey": "val",
+                    "ikey": 1
+                }]],
+                [[{
```
Comment (spacewander) [inline t/control/plugin-metadata.t:34]: We don't need to check the response of admin-api in control-api's test? This response is already checked in admin-api's test.

File: t/control/plugin-metadata.t (hunk 4 above)
Comment (spacewander) [reply t/control/plugin-metadata.t:56]: @kingluo Please update the test

File: t/control/plugin-metadata.t (hunk 5 above)
Comment (kingluo) [reply t/control/plugin-metadata.t:34]: > We don't need to check the response of admin-api in control-api's test? This response is already checked in admin-api's test.

So just check the status code?

Comment (spacewander) [reply t/control/plugin-metadata.t:34]: Yes

File: t/control/plugin-metadata.t [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                    "node": {
+                        "value": {
+                            "skey": "val",
+                            "ikey": 1
+                        }
+                    },
+                    "action": "set"
+                }]]
+                )
+            if code ~= 200 and code ~= 201 then
```
Comment (spacewander) [inline t/control/plugin-metadata.t:44]: We can use this way like the other tests?
https://github.com/apache/apisix/blob/44d20c391767d3d42c5111ea7ef6318e84633996/t/plugin/csrf.t#L83

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**